spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java

```diff
@@ -0,0 +1,490 @@
+
+import org.junit.Ignore;
+import org.junit.Test;
+import scala.Some;
+
+class WriteIntegrationTestBase extends SparkBigQueryIntegrationTestBase {
+
+  private static final String TEMPORARY_GCS_BUCKET_ENV_VARIABLE = "TEMPORARY_GCS_BUCKET";
+  private static final String LIBRARIES_PROJECTS_TABLE = "bigquery-public-data.libraries_io.projects";
+  private static final String ALL_TYPES_TABLE_NAME = "all_types";
+  protected static AtomicInteger id = new AtomicInteger(0);
+  protected BigQuery bq;
+
+
+  protected String temporaryGcsBucket = Preconditions.checkNotNull(
+      System.getenv(TEMPORARY_GCS_BUCKET_ENV_VARIABLE),
+      "Please set the %s env variable to point to a write enabled GCS bucket",
+      TEMPORARY_GCS_BUCKET_ENV_VARIABLE);
+
+  public WriteIntegrationTestBase() {
+    super();
+    this.bq = BigQueryOptions.getDefaultInstance().getService();
+  }
+
+  private Metadata metadata(String key, String value) {
+    return metadata(ImmutableMap.of(key, value));
+  }
+
+  private Metadata metadata(Map<String, String> map) {
+    MetadataBuilder metadata = new MetadataBuilder();
+    map.forEach((key, value) -> metadata.putString(key, value));
+    return metadata.build();
+  }
+
+  @Before
+  public void createTestTableName() {
+    // have a fresh table for each test
+    this.testTable = "test_" + System.nanoTime();
+  }
+
+  // Write tests. We have four save modes: Append, ErrorIfExists, Ignore and
+  // Overwrite. For each there are two behaviours - the table exists or not.
+  // See more at http://spark.apache.org/docs/2.3.2/api/java/org/apache/spark/sql/SaveMode.html
+
+  protected Dataset<Row> initialData() {
+    return spark.createDataset(Arrays.asList( //
+        new Person("Abc", Arrays.asList(
+            new Friend(10, Arrays.asList(new Link("www.abc.com"))))), //
+        new Person("Def", Arrays.asList(new Friend(12, Arrays.asList(new Link("www.def.com")))))),
+        Encoders.bean(Person.class)).toDF();
+  }
+
+  protected Dataset<Row> additonalData() {
+    return spark.createDataset(Arrays.asList( //
+        new Person("Xyz",
+            Arrays.asList(new Friend(10, Arrays.asList(new Link("www.xyz.com"))))),
+        new Person("Pqr",
+            Arrays.asList(new Friend(12, Arrays.asList(new Link("www.pqr.com")))))),
+        Encoders.bean(Person.class)).toDF();
+  }
+
+  // getNumRows returns BigInteger, and it messes up the matchers
+  protected int testTableNumberOfRows() {
+    return bq.getTable(testDataset.toString(), testTable).getNumRows().intValue();
+  }
+
+  private StandardTableDefinition testPartitionedTableDefinition() {
+    return bq.getTable(testDataset.toString(), testTable + "_partitioned")
+        .getDefinition();
+  }
+
+  protected void writeToBigQuery(Dataset<Row> df, SaveMode mode) {
+    writeToBigQuery(df, mode, "parquet");
+  }
+
+  protected void writeToBigQuery(Dataset<Row> df, SaveMode mode, String format) {
+    df.write().format("bigquery")
```

Comment:
should'nt we test for V2 as well?

Reply:
V2 will be "bigquery" as well, as it is on a separate jar